risc0/zkvm/src/host/receipt.rs

```diff
@@ -206,14 +214,29 @@ impl SegmentReceipts {
         final_receipt.verify_with_context(ctx)?;
         let metadata = final_receipt.get_metadata()?;
         log::debug!("final: {metadata:#?}");
-        if prev_image_id != metadata.pre.digest() {
+        // If there's more than one segment, the last segment could be the fault checker
+        if prev_image_id != metadata.pre.digest()
+            && !(cfg!(feature = "enable-fault-proof")
+                && receipts.len() > 0
+                && metadata.pre.digest() == FAULT_CHECKER_ID.into())
+        {
             return Err(VerificationError::ImageVerificationError);
         }
 
         if metadata.exit_code == ExitCode::SystemSplit {
             return Err(VerificationError::UnexpectedExitCode);
         }
 
+        // For receipts indicating proof of fault, the guest code should post the
+        // post-image ID of the original guest code to prove that the fault checker
+        // tried to execute the next instruction from the same state of the machine.
+        if metadata.pre.digest() == FAULT_CHECKER_ID.into() {
+            let digest: Digest = from_slice(&journal.clone()).unwrap();
+            if digest != prev_image_id {
+                return Err(VerificationError::FaultStateMismatch);
+            }
+        }
```

Comment:
What happens when the fault checker exits with a state that is not `Halted(0)`?

Reply:
It should be an error. Thanks for catching this! Fixed!